File: 22.10_PS/1015_미로_탐색.py
# boj. 2178

import sys
from collections import deque

input = sys.stdin.readline
n, m = map(int, input().split())
graph = []
visited = [[0] * m for _ in range(n)]
count = 0

for _ in range(n):
    a = list(map(int, list(input().rstrip())))
    graph.append(a)

dx = [0, -1, 1, 0]
dy = [-1, 0, 0, 1]

def bfs(x, y):
    queue = deque([(x, y)])
    visited[y][x] = True

    while queue:
        x, y = queue.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < m and 0 <= ny < n and graph[ny][nx]:
                if not visited[ny][nx]:
                    queue.append((nx, ny))
                    visited[ny][nx] = visited[y][x] + 1

bfs(0, 0)
print(visited[n - 1][m - 1])

File: 22.10_PS/1015_미로_탐색.java
// boj. 2178

import java.io.*;
import java.util.*;

public class Main {
    static int n;
    static int m;
    static int[][] graph;
    static int[][] visited;
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {-1, 1, 0, 0};

    public static void main(String args[]) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        graph = new int[n][m];
        visited = new int[n][m];
        for (int i = 0; i < n; i++) {
            String[] a = br.readLine().split("");
            for (int j = 0; j < m; j++) {
                graph[i][j] = Integer.parseInt(a[j]);
            }
        }

        BFS(0, 0);

        System.out.println(visited[n-1][m-1]);

    }

    static void BFS(int x, int y){
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(x, y));
        visited[y][x] = 1;
        int nx;
        int ny;

        while (!queue.isEmpty()){
            Node nextElement = queue.poll();
            x = nextElement.x;
            y = nextElement.y;
            for (int i = 0; i < 4; i++) {
               nx = x + dx[i];
               ny = y + dy[i];

                if (0 <= nx && nx < m && 0 <= ny && ny < n) {
                    if (graph[ny][nx] == 1 && visited[ny][nx] == 0) {
                        queue.add(new Node(nx, ny));
                        visited[ny][nx] = visited[y][x] + 1;
                    }
                }
            }
        }
    }

    static class Node {
        public int x;
        public int y;

        Node(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
}
